## Откройте файл с данными и изучите общую информацию. 

Загрузим данные и посмотрим общую информацию о таблице:

In [1]:
import pandas as pd
data = pd.read_csv('data.csv')
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Посмотрим первые 10 строк таблицы:

In [2]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Посмотрим последние 10 строк таблицы:

In [3]:
data.tail(10)  


children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

Посмотрим где и сколько пропусков:

In [4]:
data.isna().sum() 


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Проверим данные о количестве детей:

In [5]:
data['children'].value_counts() 

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Узнаем минимальный трудовой стаж в днях:

In [6]:
data['days_employed'].min() 

-18388.949900568383

Узнаем максимальный трудовой стаж в днях:

In [7]:
data['days_employed'].max() 

401755.40047533

Узнаем минимальныый возраст окредитованных:

In [8]:
data['dob_years'].min() 

0

Узнаем максимальныый возраст окредитованных:

In [9]:
data['dob_years'].max() 

75

Проверим данные о образовании:

In [10]:
data['education'].value_counts() 

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Проверим количество групп в образовании:

In [11]:
data['education_id'].value_counts() 

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

Проверим данные о семейном положении:

In [12]:
data['family_status'].value_counts() 

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Проверим соответсвуют ли данные числовых обозначений семейного положения:

In [13]:
data['family_status_id'].value_counts() 

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Узнаем сколько в данных мужчин и женщин:

In [14]:
data['gender'].value_counts() 

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Посмотрим данные с гендером 'XNA':

In [15]:
data[data['gender']=='XNA'] 

children  days_employed  dob_years            education  education_id  \
10701         0   -2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Посмотрим на тип занятости кредитующихся:

In [16]:
data['income_type'].value_counts() 

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

Сколько всего имеют задолжность по кредиту,а сколько нет:

In [17]:
data['debt'].value_counts() 

0    19784
1     1741
Name: debt, dtype: int64

Уровень максимального дохода:

In [18]:
data['total_income'].max() 

2265604.028722744

Уровень минимального дохода:

In [19]:
data['total_income'].min() 

20667.26379327158

На какие цели люди берут кредит:

In [20]:
data['purpose'].unique() 

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

### Вывод

При просмотре общих сведений в данных были обнаружены, пропуски в столбцах: days_employed и total_income.
*days_employed — общий трудовой стаж в днях.
*total_income — ежемесячный доход.
Строки с пропущенными значениями совпадают, скорей всего это ошибка выгрузки.
Пропущенны количественные данные.Тип данных float 64. Всего пропущено 2174 из 21525 что составляет 10 % но так как они встречаются в столбцах не нужных для исследования, можно их пропустить и заменить на 0.
В столбце ['days_employed'] - указаны значения точно не в днях, так как он не нужен для дальнейших вычислений им можно пренебречь, предположительно по типу данных это что то похожее на даты в ситеме UNIX с каго то определенного дня( так как при выполнении 'data.groupby('income_type')['days_employed'].median()' сильно отличаются показатели у безработных и пенсионеров)
В столбце['children'] есть отрицательные значения чего по характеру данных быть не может нужно исправить,так же в столбце ['children'] встречается чило 20 76 раз скорей всего это ошибка ввода так как 2 и 0 расположены рядом число не большое и можно пренебречь, но лучше исправить на 2.
В столбце ['dob_years'] - есть люди с возрастом 0 их лучше заменить на медину

## Предобработка данных

### Обработка пропусков

Заменим NaN на 0 в уровне дохода(потом 0 заменим средним, так как я считаю что с нулевым доходом кредит выдать не могли и 0 это тоже пропуск) и проверим:

In [21]:
data['total_income'] = data['total_income'].fillna(0)

Заменим NaN на медиану в days_employed и проверим:

In [22]:
days_employed_median = data['days_employed'].median()
data['days_employed'] = data['days_employed'].fillna(days_employed_median)
data.isna().sum() 

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Уберем отрицательные значения в столбце с количеством детей и проверим:

In [23]:
data['children'] = data['children'].abs() 
data['children'].value_counts() 


0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

Заменим 20 детей на 2 и проерим, так как я считаю сто это ошибка ввода:

In [24]:
data.loc[data['children'] == 20, 'children'] = 2 
data['children'].value_counts() 

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Заменим потеряный гендер на 'F' так как женщин больше:

In [25]:
data.loc[data['gender'] == 'XNA', 'gender'] = 'F' 
data['gender'].value_counts() 

F    14237
M     7288
Name: gender, dtype: int64

Заменим пропущенные значения в возрасте на медиану, и проверим что получилось:

In [26]:
dob_years_median = data['dob_years'].median()
data.loc[data['dob_years'] == 0, 'dob_years'] = dob_years_median 
data['dob_years'].min() 


19

Я считаю что с нулевым доходом кредит бы не одобрили по этому хочу заменить уровень дохода равный 0 на медиану:

In [27]:
total_income_median = data['total_income'].median()
data.loc[data['total_income'] == 0.0, 'total_income'] = total_income_median 
data['total_income'].min() 


20667.26379327158

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Вывод

В этом разделе мы мзбавили таблицу от пропусков. 
Нули в  возрасте заменили медианой, так же медианой заменили 0 появившиеся после замены NaN в уровене дохода. 
Избавились от отрицательных значения в столбце с количеством детей.

## Замена типа данных

Заменим уровень дохода на целочисленные так удобней изучать и проверим:

In [29]:
data['total_income'] = data['total_income'].astype('int') 
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


Заменим возраст на целочисленные, возраст указывается по количеству полных лет и проверим:

In [30]:
data['dob_years'] = data['dob_years'].astype('int') 
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


### Вывод

Заменив тип данных в столбце дохода и возраста на целочисленные мы сделали таблицу визуально удобней.
Был выбран метод astype('int') потому что нам нужны были имено целочисленные значения('integer').

## Обработка дубликатов

Посмотрим какие у нас есть уникальные значения в столбце образование:

In [31]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Приведем все значения столбце образование в нижний регистр:

In [32]:
data['education'] = data['education'].str.lower() 

Узнаем количество дубликатов:

In [33]:
data.duplicated().sum() 

72

Посмотрим что это за дубликаты:

In [34]:
data[data.duplicated()].head() 

children  days_employed  dob_years education  education_id  \
2849         0   -1203.369529         41   среднее             1   
3290         0   -1203.369529         58   среднее             1   
4182         1   -1203.369529         34    высшее             0   
4851         0   -1203.369529         60   среднее             1   
5557         0   -1203.369529         58   среднее             1   

         family_status  family_status_id gender income_type  debt  \
2849   женат / замужем                 0      F   сотрудник     0   
3290  гражданский брак                 1      F   пенсионер     0   
4182  гражданский брак                 1      F   сотрудник     0   
4851  гражданский брак                 1      F   пенсионер     0   
5557  гражданский брак                 1      F   пенсионер     0   

      total_income                  purpose  
2849        135514  покупка жилья для семьи  
3290        135514          сыграть свадьбу  
4182        135514                  свадьба  
4851        135514                  свадьба  
5557        135514          сыграть свадьбу

Удалим все дубликаты в таблице и проверим:

In [35]:
data = data.drop_duplicates().reset_index(drop= True) 
data.duplicated().sum() 

0

### Вывод

В данном разделе мы привели в нижний регистр столбец с образованием, после чего проверили колличество дубликатов в таблице, судя по тому как это выглядит большая часть этих дубликатов появилась после замены NaN, но так как мы обнаружили всего 72 записи что не составляет менее половины процента от общей массы данных, данные дубликаты можно удалить и не тратить на них время. Так же в следующим шагом мы нормализуем столбец purpose и сможем ещё подчистить таблицу от дубликатов.

## Лемматизация

Импортируем нужную библиотеку,проведём лемматизацию и выведем результат на экран:

In [36]:
from pymystem3 import Mystem 
m = Mystem()
target = data['purpose'].apply(m.lemmatize) 
target.value_counts() 

Installing mystem to /home/timur/.local/bin/mystem from http://download.cdn.yandex.net/mystem/mystem-3.1-linux-64bit.tar.gz


TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[автомобиль, \n]                                          972
[свадьба, \n]                                             791
[на,  , проведение,  , свадьба, \n]                       767
[сыграть,  , свадьба, \n]                                 765
[операция,  , с,  , недвижимость, \n]                     675
[покупка,  , коммерческий,  , недвижимость, \n]           661
[операция,  , с,  , жилье, \n]                            652
[покупка,  , жилье,  , для,  , сдача, \n]                 651
[операция,  , с,  , коммерческий,  , недвижимость, \n]    650
[покупка,  , жилье, \n]                                   646
[жилье, \n]                                               646
[покупка,  , жилье,  , для,  , семья, \n]                 638
[строительство,  , собственный,  , недвижимость, \n]      635
[недвижимость, \n]                                        633
[операция,  , со,  , свой,  , недвижимость, \n]           627
[строительство,  , жилой,  , недвижимость, \n]            624
[покупка

Напишем функцию которая приведет все знеачения к 4 вариантам, применим эту функцию к столбцу целей 
и создадим новый более понятный:

In [37]:
def credit_target(target):       
    if 'образование' in target:
            return 'Образование'
    if 'свадьба' in target:
            return 'Свадьба'
    if 'автомобиль' in target:
            return 'Автомобиль'    
    if 'недвижимость'or 'жилье' in target:
            return 'Недвижимость'             
        
data['target'] = data['purpose'].apply(credit_target) 
data['target'].value_counts() 

Недвижимость    16991
Образование      2699
Автомобиль        972
Свадьба           791
Name: target, dtype: int64

Проверяю на дубликаты ещё раз после приведения целей в порядок

In [38]:
data.duplicated().sum() 

0

### Вывод

В столбце ['purpose'] у нас много уникальных значений при чем некоторые из них означают одно и тоже что бы легче было изучать данные сократим варианты.Проведем лематизацию в столбце ['purpose'] Посмотрим какие там уникальные значения встречаются, после чего напишем функцию которая вернет нам всего 4 значения.Сохраним их в столбце ['target'] и проверим что получилось. Теперь у нас есть отдельный столбец с целями по которому будет удобно сортировать таблицу.

## Категоризация данных

Посмотрим что у нас  доходами по типу занятости:

In [39]:
data.groupby('income_type').agg({'purpose': 'count','total_income':'mean'}) 



purpose   total_income
income_type                            
безработный            2  131339.000000
в декрете              1   53829.000000
госслужащий         1457  167376.417982
компаньон           5078  195816.261914
пенсионер           3829  136964.367981
предприниматель        2  317338.500000
сотрудник          11083  158884.899756
студент                1   98201.000000

Как то не очень наглядно, было бы лучше если бы было разделение дохода на: ниже среднего, средний и выше среднего.

Определим значения квантелей:

In [40]:
total_income = data['total_income'].tolist()#Создаем переменную в виде списка зарплат
income_list = pd.qcut(total_income, 4) #Передаем список функции qcut и просим разделить на 4 части 
income_list

[(195818.0, 2265604.0], (107620.0, 135789.0], (135789.0, 195818.0], (195818.0, 2265604.0], (135789.0, 195818.0], ..., (195818.0, 2265604.0], (135789.0, 195818.0], (20666.999, 107620.0], (195818.0, 2265604.0], (20666.999, 107620.0]]
Length: 21453
Categories (4, interval[float64]): [(20666.999, 107620.0] < (107620.0, 135789.0] < (135789.0, 195818.0] < (195818.0, 2265604.0]]

Обозначим границы дохода на основе значений квантелей (возможно данные будут немного отличаться от офицальных так как по данным РБК на 2018 год средний доход был 72,7 - 121,1 (https://www.rbc.ru/economics/18/06/2018/5b27bf829a794710796bbfe5) Но во всяком случае в средний класс попадет 50% выборки

In [41]:
medium = 107798.0 
high = 195543.0 

Переменные обозначают границы разделения дохода.Напишем функцию которая будет возвращать оценку дохода: ниже среднего, средний и выше среднего.

In [42]:
total_= data['total_income'] #Обозначаем переменную с зарплатами.

def income(total):
    if total < medium:
        return 'ниже среднего'
    if (total >= medium)&(total < high):
        return 'средний'
    else:
        return 'высокий'
    
data['income_status'] = data['total_income'].apply(income) #Применяем эту функию к нужному столбцу и создаем новый
data['income_status'].value_counts() #Проверяем


средний          10690
высокий           5382
ниже среднего     5381
Name: income_status, dtype: int64

In [43]:
data.columns #Для того что бы посмотреть как котегоризировать

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose', 'target', 'income_status'],
      dtype='object')

Словарь для семйного положения:

In [44]:
family_dict = data[['family_status_id', 'family_status']] #Соберем словарь
family_dict = family_dict.drop_duplicates().reset_index(drop = True)
family_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Словарь уровня образовагия:

In [45]:
education_dict = data[['education_id', 'education']]
education_dict = education_dict.drop_duplicates().reset_index(drop = True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

В таблице есть данные о типе занятости из них тоже можо собрать словарь нужно толькло создать столбец ['income_type_id']

Просмотрим уникальные значения в столбце ['income_type_id']

In [46]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

Пишем функцию которая категоризирует значения типов занятости:

In [47]:
type_id = data['income_type']

def sort(type_id):
    if 'сотрудник' in type_id:
            return 1
    if 'пенсионер' in type_id:
            return 2
    if 'компаньон' in type_id:
            return 3   
    if 'госслужащий' in type_id:
            return 4 
    if 'безработный' in type_id:
            return 5
    if 'предприниматель' in type_id:
            return 6
    if 'студент' in type_id:
            return 7   
    if 'в декрете' in type_id:
            return 0 

data['income_type_id'] = data['income_type'].apply(sort)    
data['income_type_id'].value_counts() #Проверяем


1    11083
3     5078
2     3829
4     1457
6        2
5        2
7        1
0        1
Name: income_type_id, dtype: int64

Теперь можем собрать новый словарь income_type_dict:

In [48]:
income_type_dict = data[['income_type_id', 'income_type' ]]
income_type_dict = income_type_dict.drop_duplicates().reset_index(drop = True)
income_type_dict

income_type_id      income_type
0               1        сотрудник
1               2        пенсионер
2               3        компаньон
3               4      госслужащий
4               5      безработный
5               6  предприниматель
6               7          студент
7               0        в декрете

Таким же способом можем создать столбец ['target_id']

In [49]:
target_id = data['target']

def credit_target_id(target_id): 
    if 'Образование' in target_id:
            return 0
    if 'Свадьба' in target_id:
            return 1
    if 'Автомобиль' in target_id:
            return 2   
    if 'Недвижимость' in target_id:
            return 3
data['target_id'] = data['target'].apply(credit_target_id)
data['target_id'].value_counts()
    

3    16991
0     2699
2      972
1      791
Name: target_id, dtype: int64

Создадим новый словарь tsrget_dict

In [50]:
target_dict = data[['target_id','target']]
target_dict = target_dict.drop_duplicates().reset_index(drop = True)
target_dict

target_id        target
0          3  Недвижимость
1          0   Образование
2          2    Автомобиль
3          1       Свадьба

Таким же способом можем создать столбец ['income_status_id']

In [51]:
income_name = data['income_status']

def income_id(income_name): 
    if 'средний' in income_name:
            return 0
    if 'высокий' in income_name:
            return 1
    if 'ниже среднего' in income_name:
            return 2  
    
data['income_status_id'] = data['income_status'].apply(income_id)
data['income_status_id'].value_counts()
    

0    10690
1     5382
2     5381
Name: income_status_id, dtype: int64

Создадим новый словарь income_status_dict

In [52]:
income_status_dict = data[['income_status_id','income_status']]
income_status_dict = income_status_dict.drop_duplicates().reset_index(drop = True)
income_status_dict

income_status_id  income_status
0                 1        высокий
1                 0        средний
2                 2  ниже среднего

Посмотрим каие у нас есть столбцы и оставим в таблице только нужное:

In [53]:
data.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose', 'target', 'income_status', 'income_type_id',
       'target_id', 'income_status_id'],
      dtype='object')

In [54]:
data_sort = data[['children', 'dob_years','education_id','family_status_id', 'gender', 'income_type_id', 
                  'debt', 'income_status_id']]
data_sort.head()

children  dob_years  education_id  family_status_id gender  income_type_id  \
0         1         42             0                 0      F               1   
1         1         36             1                 0      F               1   
2         0         33             1                 0      M               1   
3         3         32             1                 0      M               1   
4         0         53             1                 1      F               2   

   debt  income_status_id  
0     0                 1  
1     0                 0  
2     0                 0  
3     0                 1  
4     0                 0

In [55]:
data_sort.info() #Посмотрим информацию о нашей таблице

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   dob_years         21453 non-null  int64 
 2   education_id      21453 non-null  int64 
 3   family_status_id  21453 non-null  int64 
 4   gender            21453 non-null  object
 5   income_type_id    21453 non-null  int64 
 6   debt              21453 non-null  int64 
 7   income_status_id  21453 non-null  int64 
dtypes: int64(7), object(1)
memory usage: 1.3+ MB


Так не слишком наглядно зато таблица использует меньше памяти и так её удобно хранить. Для расшифровки данных
мы созданы следующие словари family_dict, education_dict, income_type_dict, target_dict, income_status_dict

In [56]:
family_pivot = data_sort.pivot_table(index=['family_status_id'], values='debt', 
                                    aggfunc='mean')
                           
family_pivot  #Посмотрим как влияет на возврат кредита семейное положение

debt
family_status_id          
0                 0.075452
1                 0.093494
2                 0.065693
3                 0.071130
4                 0.097509

In [57]:
income_pivot = data_sort.pivot_table(index=['income_status_id'], values='debt', 
                                    aggfunc='mean')
income_pivot #Посмотрим как доход влияет на возврат кредита

debt
income_status_id          
0                 0.086810
1                 0.071721
2                 0.079353

### Вывод

В этом разделе мы добавили новую категорию в таблицу, взяв данные дохода и разделив из на уровни: ниже среднего, средний и высокий. Также категоризировали данные оьразования, семейного положения, уровня дохода и целей кредита.Создали для этих категорий соответсвуюшие словари. Так же создали сводные таблицы по котрым легко ответить на вопросы данного проекта.

## Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [58]:
data.groupby('children').agg({'purpose': 'count', 'debt': 'mean'})

purpose      debt
children                   
0           14090  0.075444
1            4855  0.091658
2            2128  0.094925
3             330  0.081818
4              41  0.097561
5               9  0.000000

### Ответ

Из таблице Выше видно что отсутвие детей хорошо влияет на возврат кредита
<a id="1"></a>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [59]:
data.groupby('family_status').agg({'purpose': 'count', 'debt': 'mean'})

purpose      debt
family_status                           
Не женат / не замужем     2810  0.097509
в разводе                 1195  0.071130
вдовец / вдова             959  0.065693
гражданский брак          4150  0.093494
женат / замужем          12339  0.075452

### Ответ

В таблице выше видно что лучше всего долги возвращают вдовцы разведенные и люди сосотоящие в браке
<a id="2"></a>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [60]:
data.groupby('income_status').agg({'purpose': 'count', 'debt': 'mean'})

purpose      debt
income_status                   
высокий           5382  0.071721
ниже среднего     5381  0.079353
средний          10690  0.086810

### Ответ

Из таблицы видно что в чаще всего не возвращают кредит люди со средним доходом, возможно это связано с тем что они чаще чем все остальные берут кредиты.
<a id="3"></a>

- Как разные цели кредита влияют на его возврат в срок?

In [61]:
data.groupby('target').agg({'purpose': 'count', 'debt': 'mean'})

purpose      debt
target                         
Автомобиль        972  0.092593
Недвижимость    16991  0.079513
Образование      2699  0.087440
Свадьба           791  0.080910

### Ответ

Из таблицы видно что лучше всего справляются с кредитом люди взявшие кредит на недвижимость, хуже те кто взял кредит на автомобиль
<a id="4"></a>

## Общий вывод

Изучив предоставленные данные удалось выявить выявить ответы на поставленные вопросы а именно:
* Отсутвие детей хорошо влияет на возврат кредита
* Лучше всего долги возвращают вдовцы разведенные и люди сотоящие в браке.
* Чаще всего не возвращают кредит люди со средним доходом,возможно это связано с тем что они чаще остальных груп берут кредиты.
* Лучше всего справляются с кредитом люди взявшие кредит на недвижимость, хуже те кто взял кредит на автомобиль.
Рекомендации: Нужно прояснить данные по общему трудовому стажу в днях, проверять процес выгрузки так как у нас потерялось 10 % данных. Стоит обратить внимание на анамальные значения в колличестве детей или это ошибка ввода, а так же 0 значения в возрасте и уровне дохода чего точно быть не могло.